win32/tests/file_win32_ut/file_win32_ut.c

```diff
@@ -0,0 +1,952 @@
+    PTP_CALLBACK_ENVIRON captured_pcbe;
+    PTP_CLEANUP_GROUP ptp_cleanup_group = (PTP_CLEANUP_GROUP)44;
+    PTP_IO ptp_io = (PTP_IO)45;
+
+
+    STRICT_EXPECTED_CALL(malloc(IGNORED_ARG));
+    STRICT_EXPECTED_CALL(mock_CreateFileA(filename, GENERIC_READ | GENERIC_WRITE, FILE_SHARE_READ, NULL, OPEN_ALWAYS, FILE_FLAG_OVERLAPPED | FILE_FLAG_WRITE_THROUGH, NULL))
+        .SetReturn(handle);
+    STRICT_EXPECTED_CALL(mock_SetFileCompletionNotificationModes(handle, FILE_SKIP_COMPLETION_PORT_ON_SUCCESS))
+        .SetReturn(TRUE);
+    STRICT_EXPECTED_CALL(mock_InitializeThreadpoolEnvironment(IGNORED_ARG))
+        .CaptureArgumentValue_pcbe(&captured_pcbe);
+    STRICT_EXPECTED_CALL(execution_engine_win32_get_threadpool(execution_engine))
+        .SetReturn(ptp_pool);
+    STRICT_EXPECTED_CALL(mock_SetThreadpoolCallbackPool(IGNORED_ARG, ptp_pool))
+        .ValidateArgumentValue_pcbe(&captured_pcbe);
+    STRICT_EXPECTED_CALL(mock_CreateThreadpoolCleanupGroup())
+        .SetReturn(ptp_cleanup_group);
+    STRICT_EXPECTED_CALL(mock_SetThreadpoolCallbackCleanupGroup(IGNORED_ARG, ptp_cleanup_group, IGNORED_ARG));
+    STRICT_EXPECTED_CALL(mock_CreateThreadpoolIo(handle, IGNORED_ARG, NULL, IGNORED_ARG))
+        .CaptureArgumentValue_pcbe(&captured_pcbe)
+        .CaptureArgumentValue_pfnio(captured_callback)
+        .SetReturn(ptp_io);
+
+    ///act
+    FILE_HANDLE file_handle = file_create(execution_engine, filename, NULL, NULL);
+
+    ///assert
+    ASSERT_IS_NOT_NULL(file_handle);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    umock_c_reset_all_calls();
+    return file_handle;
+}
+
+static void get_callback(FILE_CB user_callback, void* user_context, uint32_t size, PTP_WIN32_IO_CALLBACK* captured_callback, LPOVERLAPPED* captured_ov)
+{
+    ///arrange
+    FILE_HANDLE file_handle = get_file_handle("test_file.txt", captured_callback);
+    unsigned char buf[10];
+    HANDLE h_event = (HANDLE)44;
+    uint64_t position = 5;
+    PTP_IO captured_ptp_io;
+
+    STRICT_EXPECTED_CALL(malloc(IGNORED_ARG));
+    STRICT_EXPECTED_CALL(mock_CreateEvent(IGNORED_ARG, IGNORED_ARG, IGNORED_ARG, IGNORED_ARG))
+        .SetReturn(h_event);
+    STRICT_EXPECTED_CALL(mock_StartThreadpoolIo(IGNORED_ARG))
+        .CaptureArgumentValue_pio(&captured_ptp_io);
+
+    STRICT_EXPECTED_CALL(mock_WriteFile(IGNORED_ARG, buf, size, NULL, IGNORED_ARG))
+        .CaptureArgumentValue_lpOverlapped(captured_ov)
+        .SetReturn(FALSE);
+    STRICT_EXPECTED_CALL(mock_GetLastError())
+        .SetReturn(ERROR_IO_PENDING);
+
+    ///act
+    FILE_WRITE_ASYNC_RESULT result = file_write_async(file_handle, buf, size, position, user_callback, user_context);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_OK, result);
+    
+    //cleanup
+    file_destroy(file_handle);

```

Comment:
>file_destroy(file_handle); [](start = 4, length = 26)

We shouldn't destroy this if you are going to call the callback after this function returns (which I assume you are doing based on the name of this function, but I haven't gotten that far).